From b2fe7fb0eb3af031f8de96e3a6934589d40b10e2 Mon Sep 17 00:00:00 2001
From: lukasloetkolben <61192133+lukasloetkolben@users.noreply.github.com>
Date: Sat, 11 Apr 2026 11:43:53 +0200
Subject: [PATCH] xnor-tech: F4 panda support, Tesla/VW MEB ignition detection

- Add F4 device types (WHITE, BLACK) to supported devices
- Add VW MEB ignition detection via CAN address 0x3C0
- Add Tesla Model S legacy ignition detection via GTW_status (0x348)
- Add deprecated HW warning for F4 pandas
---
 board/drivers/can_common.h | 21 +++++++++++++++++++++
 python/__init__.py         | 14 ++++++++++----
 python/constants.py        | 16 ++++++++++++++++
 3 files changed, 47 insertions(+), 4 deletions(-)

diff --git a/board/drivers/can_common.h b/board/drivers/can_common.h
index bbb3bc92..09712056 100644
--- a/board/drivers/can_common.h
+++ b/board/drivers/can_common.h
@@ -156,6 +156,8 @@ void can_set_forwarding(uint8_t from, uint8_t to) {
 #endif
 
 void ignition_can_hook(CANPacket_t *msg) {
+  int len = GET_LEN(msg);
+
   if (msg->bus == 0U) {
     int len = GET_LEN(msg);
 
@@ -200,6 +202,25 @@ void ignition_can_hook(CANPacket_t *msg) {
       ignition_can = (msg->data[0] >> 5) == 0x6U;
       ignition_can_cnt = 0U;
     }
+
+    // Volkswagen MEB exception
+    if ((msg->addr == 0x3C0U) && (len == 4)) {
+      ignition_can = GET_BIT(msg, 17U);
+      ignition_can_cnt = 0U;
+    }
+  }
+
+  // Tesla Model S exception
+  if (((msg->bus == 0U) || (msg->bus == 1U)) && (msg->addr == 0x348U) && (len == 8)) {
+    int counter = msg->data[6] & 0xFU;
+
+    static int prev_counter_tesla_legacy = -1;
+    if ((counter == ((prev_counter_tesla_legacy + 1) % 16)) && (prev_counter_tesla_legacy != -1)) {
+      // GTW_status
+      ignition_can = (msg->data[0] & 0x1U) != 0U;
+      ignition_can_cnt = 0U;
+    }
+    prev_counter_tesla_legacy = counter;
   }
 
   // body v2 exception
diff --git a/python/__init__.py b/python/__init__.py
index 4c954b23..dd5950fd 100644
--- a/python/__init__.py
+++ b/python/__init__.py
@@ -131,6 +131,8 @@ class Panda:
 
   # from https://github.com/commaai/openpilot/blob/103b4df18cbc38f4129555ab8b15824d1a672bdf/cereal/log.capnp#L648
   HW_TYPE_UNKNOWN = b'\x00'
+  HW_TYPE_WHITE = b'\x01'
+  HW_TYPE_BLACK = b'\x03'
   HW_TYPE_RED_PANDA = b'\x07'
   HW_TYPE_TRES = b'\x09'
   HW_TYPE_CUATRO = b'\x0a'
@@ -141,6 +143,7 @@ class Panda:
   HEALTH_STRUCT = _parse_c_struct(os.path.join(BASEDIR, "board/health.h"), "health_t")
   CAN_HEALTH_STRUCT = struct.Struct("<BIBBBBBBBBIIIIIIIHHBBBIIII")
 
+  F4_DEVICES = [HW_TYPE_WHITE, HW_TYPE_BLACK]
   H7_DEVICES = [HW_TYPE_RED_PANDA, HW_TYPE_TRES, HW_TYPE_CUATRO, HW_TYPE_BODY]
   SUPPORTED_DEVICES = H7_DEVICES
 
@@ -222,6 +225,10 @@ class Panda:
     self.health_version, self.can_version = self.get_packets_versions()
     logger.debug("connected")
 
+    hw_type = self.get_type()
+    if hw_type not in self.SUPPORTED_DEVICES:
+      print("WARNING: Using deprecated HW")
+
     # disable openpilot's heartbeat checks
     if self._disable_checks:
       self.set_heartbeat_disabled()
@@ -348,9 +355,6 @@ class Panda:
     return []
 
   def reset(self, enter_bootstub=False, enter_bootloader=False, reconnect=True):
-    if enter_bootstub or enter_bootloader:
-      assert (hw_type := self.get_type()) in self.SUPPORTED_DEVICES, f"Unknown HW: {hw_type}"
-
     # no response is expected since it resets right away
     timeout = 5000 if isinstance(self._handle, PandaSpiHandle) else 15000
     try:
@@ -430,7 +434,9 @@ class Panda:
       pass
 
   def flash(self, fn=None, code=None, reconnect=True):
-    assert (hw_type := self.get_type()) in self.SUPPORTED_DEVICES, f"Unknown HW: {hw_type}"
+    hw_type = self.get_type()
+    if hw_type not in self.SUPPORTED_DEVICES:
+      raise RuntimeError(f"HW type {hw_type.hex()} is deprecated and can no longer be flashed.")
 
     if self.up_to_date(fn=fn):
       logger.info("flash: already up to date")
diff --git a/python/constants.py b/python/constants.py
index d69f2050..263b1f3d 100644
--- a/python/constants.py
+++ b/python/constants.py
@@ -31,6 +31,21 @@ class McuConfig(NamedTuple):
     return self.bootstub_address + sum(self.sector_sizes[:i])
 
 
+F4Config = McuConfig(
+  "STM32F4",
+  0x463,
+  [0x4000 for _ in range(4)] + [0x10000] + [0x20000 for _ in range(11)],
+  16,
+  0x1FFF7A10,
+  0x800,
+  0x1FFF79C0,
+  0x8004000,
+  "panda.bin.signed",
+  0x8000000,
+  "bootstub.panda.bin",
+)
+
+
 H7Config = McuConfig(
   "STM32H7",
   0x483,
@@ -48,6 +63,7 @@ H7Config = McuConfig(
 
 @enum.unique
 class McuType(enum.Enum):
+  F4 = F4Config
   H7 = H7Config
 
   @property
-- 
2.43.0

